src/catchup/simulation/HistoryArchiveStream.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "catchup/simulation/HistoryArchiveStream.h"
+#include "herder/TxSetFrame.h"
+#include "history/FileTransferInfo.h"
+#include "history/HistoryManager.h"
+
+namespace stellar
+{
+
+HistoryArchiveStream::HistoryArchiveStream(TmpDir const& downloadDir,
+                                           LedgerRange const& range,
+                                           HistoryManager const& hm)
+    : mDownloadDir(downloadDir), mRange(range), mHistoryManager(hm)
+{
+    if (mRange.mFirst != 1 &&
+        mHistoryManager.nextCheckpointLedger(mRange.mFirst) != mRange.mFirst)
+    {
+        uint32_t prev = mHistoryManager.prevCheckpointLedger(mRange.mFirst);
+        readHeaderHistory(std::max<uint32_t>(1, prev));
+        readTransactionHistory();
+        readResultHistory();
+        while (mHeaderHistory.header.ledgerSeq != mRange.mFirst - 1)
+        {
+            readHeaderHistory();
+            readTransactionHistory();
+            readResultHistory();
+        }
+    }
+}
+
+LedgerHeaderHistoryEntry
+HistoryArchiveStream::readHeaderHistory(uint32_t ledgerSeq)
+{
+    LedgerHeaderHistoryEntry lhhe;
+
+    if (ledgerSeq == 1 ||
+        mHistoryManager.nextCheckpointLedger(ledgerSeq) == ledgerSeq)
+    {
+        if (mHeaderStream && mHeaderStream.readOne(lhhe))
+        {
+            throw std::runtime_error(
+                "Header stream should be exhausted at end of checkpoint");
+        }
+
+        FileTransferInfo info(
+            mDownloadDir, HISTORY_FILE_TYPE_LEDGER,
+            mHistoryManager.nextCheckpointLedger(ledgerSeq + 1) - 1);
```

Comment:
you can use checkpointContainingLedger() method for that

Reply:
Done.